librad/src/meta/user.rs

```diff
@@ -33,28 +33,53 @@ use crate::{
     },
 };
 
-#[derive(Clone, Deserialize, Serialize, Debug, Default, PartialEq)]
+#[derive(Clone, Deserialize, Serialize, Debug, PartialEq)]
 pub struct UserInfo {
+    // Marker so `EntityInfo` can deserialize correctly
+    user: (),
```

Comment:
I don't get why this is needed _as well as_ the unified enum

Reply:
This is tricky.

The problem is that `EntityInfo`, `UserInfo` and `ProjectInfo` need to be serialization-compatible, because we could deserialize the same blob inther as `Entity<UserInfo,_>` or, if we know it, with its specific type.

Also in the other direction, we could have serialized it either from a the specific type or from the union type.

To achieve this I marked `EntityInfo` as `#[serde(untagged)]`, so that its branches are expanded as if the enum did not exist.
The problem is that both `User` and `Project` have fully optional fields, so it is possible that all of them are `None`. In that case serde will pick the wrong branch (Murphy's law).

Adding a mandatory `user` field to `User` and a `project` firld to `Project` makes so that serde can tell them apart.
Typing them as `()` makes them zero sized so that they take space in the blob (where they are needed as discriminaor) but not in memory.

I could not find a better way for keeping the serialized blob compatible with all the enum and specific types at the same time :smiley:

Reply:
I see... so like a manual "internally tagged". Nice.

I think we should take another pass on specifying how these documents look like, and how their schema is extensible... (#9)

Reply:
> Also in the other direction, we could have serialized it either from a the specific type or from the union type.

Ok, this is starting to sound worrying to me. Sometimes we use the enumerated entity, sometimes we use the specific entity. I thought the whole idea of the enumeration is to allow us to gather any type of entity, and then inspect at run-time.

In my mind, I would see the enumeration solving the problem by serialising _that_. Which would give us the enumeration tag when serialising.

I just feel like something isn't lining up here? :thinking: But perhaps I'm wrong :)

Reply:
> In my mind, I would see the enumeration solving the problem by serialising _that_. Which would give us the enumeration tag when serialising.

With this implementation there are only two kind of blobs: the `User` and the `Project`.
If you serialize the enum you will get the specific blob anyway, and you will be able to deserialize it with either the specific type or the enum.

This is achieved by always putting the tag in the blob in the same way.
As Kim explained, it's like a manually built "internally tagged" representation.

Reply:
Ya, so I'm understanding what the setup is here, but what I don't understand is what is it buying us? It seems like we're jumping through a lot of hoops here having two ways of expressing that we can have multiple entities, whereas we could express it one way: https://github.com/radicle-dev/radicle-link/pull/170#discussion_r439344579